// ListWatcher/User.js
var fs      = require("fs");
var path    = require("path");
var _       = require("underscore");


module.exports = {

    LISTS_FILE_DIR       : "lists",

    LISTS_FILE_EXTENSION : ".list",

    LIST_NAME_SEPARATOR  : " => ",

    create: function(id) {
        return Object.create(this).setId(id);
    },

    setId: function(id) {
        this.id = id;
        return this;
    },

    getId: function() {
        return this.id;
    },

    getListsFilePath: function() {
        var name = this.getId() + this.LISTS_FILE_EXTENSION;

        return path.join(process.cwd(), this.LISTS_FILE_DIR, name);
    },

    getLists: function(callback) {
        var self   = this,
            stream = fs.createReadStream(self.getListsFilePath()),
            data   = "";

        stream.on("data", function(chunk) {
            data += chunk;
        });

        stream.on("error", function(error) {
            if (error.code == "ENOENT") {
                return callback(undefined, false);
            }

            callback(error);
        });

        stream.on("end", function() {
            var result = {};

            _.each(data.split("\n"), function(list) {
                if (!list) {
                    return;
                }

                var parts = list.split(self.LIST_NAME_SEPARATOR);

                result[parts[0]] = parts[1];
            });

            callback(undefined, result);
        });
    },

    updateLists: function(lists, callback) {
        var self   = this,
            stream = fs.createWriteStream(self.getListsFilePath());

        stream.on("close", function() {
            callback();
        });
        stream.on("error", function(error) {
            callback(error);
        });

        _.each(lists, function(name, id) {
            stream.write(id + self.LIST_NAME_SEPARATOR + name + "\n");
        });

        stream.end();
    }

};
